runtime/nvqir/cutensornet/simulator_tensornet_register.cpp

```diff
@@ -20,20 +23,45 @@ class SimulatorTensorNet : public SimulatorTensorNetBase {
     // Note: this requires CUTENSORNET_COMM_LIB as described in
     // the Getting Started section of the cuTensorNet library documentation
     // (Installation and Compilation).
-    if (cudaq::mpi::is_initialized())
+    if (cudaq::mpi::is_initialized()) {
       initCuTensornetComm(m_cutnHandle);
+      m_cutnMpiInitialized = true;
+    }
   }
   // Nothing to do for state preparation
   virtual void prepareQubitTensorState() override {}
   virtual std::string name() const override { return "tensornet"; }
   // Add a hook to reset the cutensornet MPI Comm before MPI finalization
   // to make sure we have a clean shutdown.
   virtual void tearDownBeforeMPIFinalize() override {
-    if (cudaq::mpi::is_initialized())
+    if (cudaq::mpi::is_initialized()) {
       resetCuTensornetComm(m_cutnHandle);
+      m_cutnMpiInitialized = false;
+    }
   }
+
+private:
+  friend nvqir::CircuitSimulator * ::getCircuitSimulator_tensornet();
+  /// @brief Have we initialized cuTensorNet MPI during ctor?
+  bool m_cutnMpiInitialized = false;
 };
 } // namespace nvqir
 
 /// Register this Simulator class with NVQIR under name "tensornet"
-NVQIR_REGISTER_SIMULATOR(nvqir::SimulatorTensorNet, tensornet)
+extern "C" {
```

Comment:
Q: Why is the extern "C" needed? Python entry point?

Reply:
These are plugin-style `dlsym` hooks (`runtime/common/PluginUtils.h`) for simulators/platforms hence need to be `extern "C"`.
The runtime looks for these symbols to load the simulator/platform.